security/subsystem/src/main/java/org/jboss/as/security/service/SimpleSecurityManager.java

```diff
@@ -365,6 +365,26 @@ public void push(final String securityDomain, String userName, char[] password,
         }
     }
 
+    private void propagateSubject(final SecurityContext target, final SecurityContext source) {
+        final SecurityContextUtil previousUtil = source.getUtil();
+        final SecurityContextUtil currentUtil = target.getUtil();
+        if(target.getSecurityDomain() != null && source.getSecurityDomain() != null && target.getSecurityDomain().equals(source.getSecurityDomain())){
+            target.setSubjectInfo(source.getSubjectInfo());
+        } else {
+            currentUtil.createSubjectInfo(previousUtil.getUserPrincipal(), previousUtil.getCredential(), previousUtil.getSubject());
+        }
+        if (previousUtil.getRoles() != null) {
+            try {
+                currentUtil.setRoles((RoleGroup) previousUtil.getRoles().clone());
+            } catch (CloneNotSupportedException e) {
+                // should not happen, RoleGroup supports clone.
+                throw new RuntimeException(e);
+            }
+        }
```

Comment:
What is the reason this block is needed?  The previous implementation was not manually cloning the RoleGroup and the bug report seems to be in relation to popping the Principal.

I have always considered role propagation in PicketBox across domains to be a bit of an undefined, does it make sense to be cloning the roles across if we both have the same or different domain?

Reply:
Not all implementations of RoleGroup had it(clone) in PIcketBox. If users have their own impl, we cant be sure they have as well.

Reply:
One minor comment about this, have you tried not cloning the roles? Did you run into issues if you don't do it? Reason I'm asking is that I'm not sure this is required - any roles established by the previous context should not come into play when creating a new one and performing a new authentication. I might be mistaken and this might be required but I'm just not sure about it.

Reply:
AFAIR, yes, tried with minimal set of cloning up until it worked.